### PR TITLE
Debug deadlock develop

### PR DESCRIPTION
**Pull request by christopherlagerhult**
Same task as #1878 but moved to develop branch

### CODE REVIEW

**Comment by christopherlagerhult** [comment]
>I think it was copied from somewhere else though. Anyway, I don't know why it uses CpuTimer so should we change it?

Should that be a new PR or do I add it to this one?

**Comment by emilwestergren** [comment]
A separate PR, preferably merged before this so you can use `WallTimer` in here as well.

**Comment by christopherlagerhult** [comment]
#1890

**Comment by marcusnaslund** [comment]
Frozen awaiting release.

**Comment by marcusnaslund** [comment]
What's the status here?

**Comment by christopherlagerhult** [comment]
I believe it's done, unless @emilwestergren has some remarks

File: source/system/native/MutexUnix.ooc [hunk 1]
```diff
@@ -19,7 +19,17 @@ version(unix || apple) {
 			super()
 		}
 		lock: override func {
-			pthread_mutex_lock(this _backend&)
+			version(debugDeadlock) {
+				timeLimitMilliseconds := 1000 //Only whole seconds
```
Comment (marcusnaslund) [inline source/system/native/MutexUnix.ooc]: `//Only whole seconds` Why?

Comment (christopherlagerhult) [inline source/system/native/MutexUnix.ooc]: Well, to not overcomplicate things as we lack a `TimeSpec + operator`. If you believe we might need fractions of a second in the future I'm happy to add it.

Comment (marcusnaslund) [inline source/system/native/MutexUnix.ooc]: I don't understand what you mean. Isn't this what the `tv_sec` and `tv_nsec` variables are for? Something like
```ooc
deadline tv_sec = currentTime tv_sec + timeLimitMilliseconds / 1000
deadline tv_nsec = currentTime tv_usec * 1000 + (timeLimitMilliseconds % 1000) * 1_000_000
```

Comment (christopherlagerhult) [inline source/system/native/MutexUnix.ooc]: Yeah, and the case where where `tv_nsec > 1 seconds`
```ooc
if (deadline tv_nsec > 1_000_000_000) {
    deadline tv_nsec -= 1_000_000_000
    deadline tv_sec += 1
}
```
Just feels like it is an unnecessary feature that adds some additional rows :)

Comment (marcusnaslund) [inline source/system/native/MutexUnix.ooc]: I see. But I disagree, that's a simple if-statement, and if we're at the level of precision where the variable specifies milliseconds, not allowing for fractions of seconds is an unexpected limitation.

```
(deadline tv_sec, deadline tv_nsec) = (currentTime tv_sec + timeLimitMilliseconds / 1000, currentTime tv_usec * 1000 + (timeLimitMilliseconds % 1000) * 1_000_000)
if (deadline tv_nsec >= 1_000_000_000)
	(deadline tv_sec, deadline tv_nsec) = (deadline tv_sec + 1, deadline tv_nsec - 1_000_000_000)
```

**Comment by marcusnaslund** [comment]
I've got no more comments after that though.

File: source/system/native/MutexUnix.ooc [hunk 2]
```diff
@@ -19,7 +19,19 @@ version(unix || apple) {
...
 		lock: override func {
-			pthread_mutex_lock(this _backend&)
+			version(debugDeadlock) {
+				timeLimitMilliseconds := 1000
+				deadline: TimeSpec
+				currentTime: TimeVal
+				gettimeofday(currentTime&, null)
+				(deadline tv_sec, deadline tv_nsec) = (currentTime tv_sec + timeLimitMilliseconds / 1000, currentTime tv_usec * 1000 + (timeLimitMilliseconds % 1000) * 1_000_000)
+				if (deadline tv_nsec >= 1_000_000_000)
+					(deadline tv_sec, deadline tv_nsec) = (deadline tv_sec + 1, deadline tv_nsec - 1_000_000_000)
```
Comment (marcusnaslund) [inline source/system/native/MutexUnix.ooc]: Perfect :+1:

**Pull request merged**